mithril-common/src/chain_observer/pallas_observer.rs

```diff
@@ -0,0 +1,229 @@
+            "Unimplemented and no fallback configured for PallasObserver",
+        ))
+    }
+}
+
+#[async_trait]
+impl ChainObserver for PallasChainObserver {
+    async fn get_current_epoch(&self) -> Result<Option<Epoch>, ChainObserverError> {
+        let mut client = self
+            .new_client()
+            .await
+            .map_err(ChainObserverError::General)?;
+
+        let statequery = client.statequery();
+
+        statequery.acquire(None).await.unwrap();
+
+        let era = queries_v16::get_current_era(statequery)
+            .await
+            .map_err(|err| ChainObserverError::General(err.into()))?;
+
+        let epoch = queries_v16::get_block_epoch_number(statequery, era)
+            .await
+            .map_err(|err| ChainObserverError::General(err.into()))?;
+
+        statequery.send_release().await.unwrap();
+        statequery.send_done().await.unwrap();
+
+        client.chainsync().send_done().await.unwrap();
+        drop(client.plexer_handle);
+
+        Ok(Some(Epoch(epoch as u64)))
+    }
+
+    async fn get_current_datums(
+        &self,
+        address: &ChainAddress,
+    ) -> Result<Vec<TxDatum>, ChainObserverError> {
+        let fallback = self.get_fallback().map_err(ChainObserverError::General)?;
+        fallback.get_current_datums(address).await
+    }
+
+    async fn get_current_stake_distribution(
+        &self,
+    ) -> Result<Option<StakeDistribution>, ChainObserverError> {
+        let fallback = self.get_fallback().map_err(ChainObserverError::General)?;
+        fallback.get_current_stake_distribution().await
+    }
+
+    async fn get_current_kes_period(
+        &self,
+        opcert: &OpCert,
+    ) -> Result<Option<KESPeriod>, ChainObserverError> {
+        let fallback = self.get_fallback().map_err(ChainObserverError::General)?;
+        fallback.get_current_kes_period(opcert).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs;
+
+    use pallas_codec::utils::AnyCbor;
+    use pallas_network::miniprotocols::{
+        chainsync::{self},
+        localstate::{self, ClientQueryRequest},
+    };
+    use tokio::net::UnixListener;
+
+    use super::*;
+    use crate::CardanoNetwork;
+
+    #[tokio::test]
+    async fn get_current_epoch_with_fallback() {
+        let server = tokio::spawn({
+            async move {
+                let socket_path = Path::new("node.socket");
+                if socket_path.exists() {
+                    fs::remove_file(socket_path).unwrap();
+                }
+                let unix_listener = UnixListener::bind(socket_path).unwrap();
+
+                let mut server = pallas_network::facades::NodeServer::accept(&unix_listener, 10)
+                    .await
+                    .unwrap();
+
+                let maybe_acquire = server.statequery().recv_while_idle().await.unwrap();
+                assert!(maybe_acquire.is_some());
+                assert_eq!(*server.statequery().state(), localstate::State::Acquiring);
```

Comment:
The subject of the test is the PallasObserver, not our mock server. We don't need to assert individual states or enforce a specific order in the queries. Our mock server should just match the type of query received and return a hardcoded value (regardless of the order).

Something like:

```rust
let result = match query {
    Request:LedgerQuery(LedgerQuery::HardForkQuery(HardForkQuery::GetCurrentEra)) => AnyCbor::from_encode(4),
    Request:LedgerQuery(LedgerQuery::BlockQuery(_, BlockQuery::GetEpochNo)) => AnyCbor::from_encode([8]),
}

server.statequery().send_result(result).await.unwrap();
```

Reply:
@scarmuega thank you, done.